src/bin/commands/demux.rs

```diff
@@ -63,6 +59,43 @@ impl ReadSet {
     const COLON: u8 = b':';
     const PLUS: u8 = b'+';
 
+    /// Produces an iterator over references to the template segments stored in this ``ReadSet``.
+    fn template_segments(&self) -> Box<dyn Iterator<Item = &FastqSegment> + '_> {
+        Box::new(self.segments.iter().filter(|s| s.segment_type == SegmentType::Template))
+    }
+    /// Produces an iterator over references to the sample barcode segments stored in this
+    /// ``ReadSet``.
+    fn sample_barcode_segments(&self) -> Box<dyn Iterator<Item = &FastqSegment> + '_> {
+        Box::new(self.segments.iter().filter(|s| s.segment_type == SegmentType::SampleBarcode))
+    }
+    /// Produces an iterator over references to the molecular barcode segments stored in this
+    /// ``ReadSet``.
+    fn molecular_barcode_segments(&self) -> Box<dyn Iterator<Item = &FastqSegment> + '_> {
+        Box::new(self.segments.iter().filter(|s| s.segment_type == SegmentType::MolecularBarcode))
+    }
+
+    /// Generates the sample barcode sequence for this read set and returns it as a Vec of bytes.
+    fn sample_barcode_sequence(&self, sample_barcode_length: usize) -> Vec<u8> {
+        let mut sample_barcode = Vec::with_capacity(sample_barcode_length);
+        for segment in self.sample_barcode_segments() {
+            sample_barcode.extend(segment.seq.clone());
```

Comment:
It doesn't feel like you should have to clone the vec, since the elements are copy?